File: testdata/jennies/builders/foreign_builder/TypescriptBuilder/src/builder_pkg/somenicebuilder_builder_gen.ts
import * as cog from '../cog';
import * as some_pkg from '../some_pkg';

export class SomeNiceBuilderBuilder implements cog.Builder<some_pkg.SomeStruct> {
    private readonly internal: some_pkg.SomeStruct;

    constructor() {
        this.internal = some_pkg.defaultSomeStruct();
    }

    build(): some_pkg.SomeStruct {
        return this.internal;
    }

    title(title: string): this {
        this.internal.title = title;
        return this;
    }
}

File: testdata/jennies/builders/foreign_builder/GoBuilder/builder_pkg/somenicebuilder_builder_gen.go
package builder_pkg

import (
	cog "github.com/grafana/cog/generated/cog"
	some_pkg "github.com/grafana/cog/generated/some_pkg"
)

var _ cog.Builder[some_pkg.SomeStruct] = (*SomeNiceBuilderBuilder)(nil)

type SomeNiceBuilderBuilder struct {
    internal *some_pkg.SomeStruct
    errors cog.BuildErrors
}

func NewSomeNiceBuilderBuilder() *SomeNiceBuilderBuilder {
	resource := some_pkg.NewSomeStruct()
	builder := &SomeNiceBuilderBuilder{
		internal: resource,
		errors: make(cog.BuildErrors, 0),
	}

	return builder
}



func (builder *SomeNiceBuilderBuilder) Build() (some_pkg.SomeStruct, error) {
	if err := builder.internal.Validate(); err != nil {
		return some_pkg.SomeStruct{}, err
	}
	
	if len(builder.errors) > 0 {
	    return some_pkg.SomeStruct{}, cog.MakeBuildErrors("builder_pkg.someNiceBuilder", builder.errors)
	}

	return *builder.internal, nil
}

func (builder *SomeNiceBuilderBuilder) Title(title string) *SomeNiceBuilderBuilder {
    builder.internal.Title = title

    return builder
}

